[Bug] [MySQL-CDC] throw java.lang.ArrayIndexOutOfBoundsException

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### What happened

```java
Caused by: java.lang.ArrayIndexOutOfBoundsException
	at com.github.shyiko.mysql.binlog.io.BufferedSocketInputStream.read(BufferedSocketInputStream.java:65)
```

### SeaTunnel Version

dev

### SeaTunnel Config

```conf
-
```


### Running Command

```shell
-
```


### Error Exception

```log
Caused by: org.apache.kafka.connect.errors.ConnectException: An exception occurred in the change event producer. This connector will be stopped.
	at io.debezium.pipeline.ErrorHandler.setProducerThrowable(ErrorHandler.java:42)
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$ReaderThreadLifecycleListener.onEventDeserializationFailure(MySqlStreamingChangeEventSource.java:1207)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:965)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:599)
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:857)
	... 1 more
Caused by: io.debezium.DebeziumException
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1154)
	... 5 more
Caused by: java.lang.ArrayIndexOutOfBoundsException
	at com.github.shyiko.mysql.binlog.io.BufferedSocketInputStream.read(BufferedSocketInputStream.java:65)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readWithinBlockBoundaries(ByteArrayInputStream.java:262)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:241)
	at java.io.InputStream.skip(InputStream.java:224)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.skipToTheEndOfTheBlock(ByteArrayInputStream.java:280)
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:305)
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:232)
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$1.nextEvent(MySqlStreamingChangeEventSource.java:230)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:952)
```


### Zeta or Flink or Spark Version

_No response_

### Java or Scala Version

_No response_

### Screenshots

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
+++ b/seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
@@ -1,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.shyiko.mysql.binlog.io;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Copied from https://github.com/osheroff/mysql-binlog-connector-java project to fix
+ * https://github.com/apache/seatunnel/issues/7380
+ *
+ * <p>reference: - https://github.com/osheroff/mysql-binlog-connector-java/issues/66 -
+ * https://github.com/apache/flink-cdc/issues/460
+ */
+public class BufferedSocketInputStream extends FilterInputStream {
+
+    private byte[] buffer;
+    private int offset;
+    private int limit;
+
+    public BufferedSocketInputStream(InputStream in) {
+        this(in, 512 * 1024);
+    }
+
+    public BufferedSocketInputStream(InputStream in, int bufferSize) {
+        super(in);
+        this.buffer = new byte[bufferSize];
+    }
+
+    @Override
+    public int available() throws IOException {
+        return limit == -1 ? in.available() : limit - offset + in.available();
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (offset < limit) {
+            return buffer[offset++] & 0xff;
+        }
+        offset = 0;
+        limit = in.read(buffer, 0, buffer.length);
+        return limit != -1 ? buffer[offset++] & 0xff : -1;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        if (offset >= limit) {
+            if (len >= buffer.length) {
+                return in.read(b, off, len);
+            }
+            offset = 0;
+            limit = in.read(buffer, 0, buffer.length);
+            if (limit == -1) {
+                return limit;
+            }
+        }
+        int bytesRemainingInBuffer = Math.min(len, limit - offset);
+        System.arraycopy(buffer, offset, b, off, bytesRemainingInBuffer);
+        offset += bytesRemainingInBuffer;
+        return bytesRemainingInBuffer;
+    }
+}

--- a/seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/test/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStreamTest.java
+++ b/seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/test/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStreamTest.java
@@ -1,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.github.shyiko.mysql.binlog.io;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class BufferedSocketInputStreamTest {
+
+    @Test
+    public void testReadFromBufferedSocketInputStream() throws Exception {
+        BufferedSocketInputStream in =
+                new BufferedSocketInputStream(
+                        new ByteArrayInputStream(
+                                new byte[] {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'}));
+        byte[] buf = new byte[3];
+        assertEquals(3, in.read(buf, 0, buf.length));
+        Arrays.equals(new byte[] {'A', 'B', 'C'}, buf);
+        assertEquals(5, in.available());
+
+        assertEquals(3, in.read(buf, 0, buf.length));
+        Arrays.equals(new byte[] {'D', 'E', 'F'}, buf);
+        assertEquals(2, in.available());
+
+        assertEquals(2, in.read(buf, 0, buf.length));
+        Arrays.equals(new byte[] {'G', 'H'}, buf);
+        assertEquals(0, in.available());
+
+        // reach the end of stream normally
+        assertEquals(-1, in.read(buf, 0, buf.length));
+        assertEquals(0, in.available());
+    }
+}

